Rework dynamic forms topic for style and organization (#36465)
angular/angular: aio/content/guide/dynamic-form.md

The technique is particularly useful when you have a type of form whose content must
change frequently to meet rapidly changing business and regulatory requirements.

This guide shows you how to use `formGroup` to dynamically
render a basic form with different control types and validation.
The basic version can evolve to support a richer variety of questions, more graceful rendering, and superior user experience.

**Prerequisites**

Before reading about dynamic forms, you should have a basic understanding to the following.

* [TypeScript programming](https://www.typescriptlang.org/docs/home.html "The TypeScript language").

* Fundamental concepts of [Angular app design](guide/architecture "Introduction to Angular app-design concepts").

* Basic knowledge of [reactive forms](guide/reactive-forms "Reactive forms guide").

## Setup

Dynamic forms are based on reactive forms. To give the application access reactive forms directives, the [root module](guide/bootstrapping "Learn about bootstrapping an app from the root module.") imports `ReactiveFormsModule` from the `@angular/forms` library.

> The example used here to illustrate the dynamic-forms technique is a reactive form that you use to build an online application experience for heroes seeking employment.

Reviewer: ```suggestion
In this tutorial, you will build a form using the dynamic forms technique. You will build an online application for heroes seeking employment.
```

# Building dynamic forms

Many forms, such as questionaires, can be very similar to one another in format and intent.
To make it faster and easier to generate different versions of such a form,
you can create a *dynamic form template* based on metadata that describes the business object model.
You can then use the template to generate new forms automatically, according to changes in the data model.

The technique is particularly useful when you have a type of form whose content must
change frequently to meet rapidly changing business and regulatory requirements.
A typical use case is a questionaire. You might need to get input from users in different contexts.
The format and style of the forms a user sees should remain constant, while the actual questions you need to ask vary with the context.

In this tutorial you will build a dynamic form that presents a basic questionaire.
You will build an online application for heroes seeking employment.
The agency is constantly tinkering with the application process, but by using the dynamic form
you can create the new forms on the fly without changing the application code.

The tutorial walks you through the following steps.

> 1. Set up the data model behind the questionaire, defining some heroes and their attributes.

Reviewer 2: I noticed that the list of steps here doesn't 1:1 reflect the actual sections in the doc below. While this might be treated as a separate logical unit describing the process (vs actual step-by-step instructions), I think it'd be great to keep them in sync so that it also serves as a kind of table of contents. I understand that the steps outlined here require more content than just a few words (that works for section headers below), but may be we can rephrase them to be more aligned? I was thinking of something like this:

- **Create data models to represent form controls**
Set up the data model behind the questionaire, defining some heroes and their attributes.

- **Compose dynamic form contents**
Compose a set of question components into a dynamic-form container component

- ...

Author: The whole section is rewritten to follow the step structure - please check again.

The tutorial walks you through the following steps.

> 1. Set up a project.

Author: Good idea 👍

2. Create the form object model.

> 3. Compose the form contents with some sample data.

Reviewer 2: I'd like to propose a bit more simplified version of the list and have the following items there:

1. Enable Forms feature for a project.
2. Establish data model to represent form controls.
3. Populate the model with sample data.
4. Develop a component to create form controls dynamically.

Author: This is just what I was hoping for - it helps to know the subject matter!

</code-tabs>

{@a object-model}

## Create a form object model

A dynamic form requires an object model that can describe all scenarios needed by the form functionality.
The example hero-application form is a set of questions&mdash;that is, each control in the form must ask a question and accept an answer.

The data model for this type of form must represent a question.
The example includes the `DynamicFormQuestionComponent`, which defines a  question as the fundamental object in the model.

The following `QuestionBase` is a base class for a set of controls that can represent the question and its answer in the form.

<code-example path="dynamic-form/src/app/question-base.ts" header="src/app/question-base.ts">

</code-example>

### Define control classes

From this base, the example derives two new classes, `TextboxQuestion` and `DropdownQuestion`,
that represent different control types.
You can bind a form to these specific question types and render the appropriate controls dynamically.

> * The `TextboxQuestion` control supports multiple HTML5 types such as text, email, and url

Reviewer 2: It also feels like the `control` term here might be a bit confusing. The `TextboxQuestion` and `DropdownQuestion` classes are more like data models that would later be used to create actual form controls. In the current version of this section, the `control` term does not refer to the `TextboxQuestion` and `DropdownQuestion` classes (which makes sense to me).

> * The `TextboxQuestion` control supports multiple HTML5 types such as text, email, and url

Author: Tried to clarify, please check again

through the `type` property.

<code-example path="dynamic-form/src/app/question-textbox.ts" header="src/app/question-textbox.ts"></code-example>

* The `DropdownQuestion` control presents a list of choices in a select box.

<code-example path="dynamic-form/src/app/question-dropdown.ts" header="src/app/question-dropdown.ts"></code-example>

### Compose form groups

The dynamic form uses a service to provide a specific set of controls to be collected into a group and added to the form.
This form group consumes the metadata from the question model and allows you to specify default values and validation rules.

The example defines a `QuestionControlService` that collects a set of individual question controls into a `FormGroup`.

> The form group consumes the metadata from the question model and allows you to specify default values and validation rules.

Reviewer 2: In this section the `control` term also refers to the instances of question classes when in the current version the `metadata from the question model` is used (which is technically correct).

Author: Another try. Please check again

<code-example path="dynamic-form/src/app/question-control.service.ts" header="src/app/question-control.service.ts"></code-example>

{@a form-component}

## Compose dynamic form contents

The dynamic form itself is represented by a container component, `DynamicFormComponent`, whose contents are provided by a question component, `DynamicFormQuestionComponent`.

> The question component provides a set of questions to include in the form using the `<app-question>' tag. This component is responsible for rendering the details of each individual question based on values in the data-bound question object.

Reviewer 2: I find the current version a bit more easier to digest, i.e.:

[there is a list of questions] -> [each question is represented via `<app-question>`] -> [`<app-question>` matches the `DynamicFormQuestionComponent`] -> [`DynamicFormQuestionComponent` represents an individual question]

May be we can rephrase first sentence to be similar to the current state? The second sentence seems fine.

From this base, the example derives two new classes, `TextboxQuestion` and `DropdownQuestion`,
that represent different control types.
When you create the form template in the next step, you will instantiate these specific question types in order to render the appropriate controls dynamically.

* The `TextboxQuestion` control type presents a question and allows users to enter input.

<code-example path="dynamic-form/src/app/question-textbox.ts" header="src/app/question-textbox.ts"></code-example>

>    The `TextboxQuestion` control type will be represented in a form template using an <input> element.

Reviewer 2: It looks like `<input>` needs additional escaping, since it's actually rendered as an HTML element on a page, [see here](https://pr36465-90d2a2e.ngbuilds.io/guide/dynamic-form).

Author: oops

   The `type` attribute of the element will be defined based on the `type` field specified in the `options` argument (for example `text`, `email`, `url`).

> 

Reviewer 2: We may need to add a space in between these lines, since the next line is very close to the line above which makes is look like it's a part of the same logical unit. Do we have some examples of this layout (unordered list with example sections) in other places to see if we can use some code from there?

* The `DropdownQuestion` control presents a list of choices in a select box.

<code-example path="dynamic-form/src/app/question-dropdown.ts" header="src/app/question-dropdown.ts"></code-example>

### Compose form groups

A dynamic form uses a service to create grouped sets of input controls, based on the form model.
The following `QuestionControlService` collects a set of `FormGroup` instances that consume the metadata from the question model. You can specify default values and validation rules.

<code-example path="dynamic-form/src/app/question-control.service.ts" header="src/app/question-control.service.ts"></code-example>

{@a form-component}

## Compose dynamic form contents

The dynamic form itself will be represented by a container component, which you will add in a later step.
Each question is represented in the form component's template by an `<app-question>` tag, which matches an instance of `DynamicFormQuestionComponent`.

The `DynamicFormQuestionComponent` is responsible for rendering the details of an individual question based on values in the data-bound question object.

> The form relies on a `formGroup` to connect the template HTML to the

Reviewer 2: Let's may be also add a link to a corresponding directive here:
```
The form relies on a [`formGroup`](api/forms/FormGroupDirective "FormGroupDirective") directive ...
```

</code-tabs>

> The `DynamicFormQuestionComponent` component can present any type of question in your model.

Reviewer 2: I'd propose to update this sentence to something like `The goal of the DynamicFormQuestionComponent is to present question types defined in your model.`.

You only have two types of questions at this point but you can imagine many more.
The `ngSwitch` statement in the template determines which type of question to display.

> The switch uses the directives `formControlName` and `formGroup` that are defined in `ReactiveFormsModule`.

Reviewer 2: ```
The switch uses directives with the `formControlName` and `formGroup` selectors. Both directives are defined in `ReactiveFormsModule`.
```
(and may be also include links to these directives?)

{@a questionnaire-data}

### Supply data

> `DynamicFormComponent` expects a list of questions in the form of an array bound to `@Input() questions`.

Reviewer 2: We refer to the `DynamicFormComponent` here, but it was not defined/discussed yet. May be we can rephrase that to say that in this section we prepare a service that would take care of supplying data. For the purposes of this guide we'll use hard coded set of questions, but in a real world app this service might fetch data from the backend/storage.

The `QuestionService` supplies this array of questions.
In a real app the service would retrieve the questions from storage, but here some data is hard coded for purposes of illustration.

The key point is that you control the hero job-application questions entirely through the objects returned from `QuestionService`.
To maintain the questionnaire maintenance as requirements change, you only need to add, update, and remove objects from the `questions` array.

<code-example path="dynamic-form/src/app/question.service.ts" header="src/app/question.service.ts">

</code-example>

{@a dynamic-template}

## Create a dynamic form template

`DynamicFormComponent` is the entry point and the main container for the form.

> The selector, `<app-dynamic-form>`, places the form in the template.

Reviewer 2: I'd propose updating it to say something like:
```
The `DynamicFormComponent` component is the entry point and the main container for the form, which is represented using the `<app-dynamic-form>` in a template.
```
